### PR TITLE
Optimize Pipeline & Add Texture Mapping

### PR DESCRIPTION
**Pull request by ZzzhHe**
# Description
This pull request includes several key enhancements:

+ **Optimized Triangle Drawing Algorithm**: Improved performance in triangle rasterization.
+ **Refactored Shader Structure**: Restructured shaders for better clarity and maintainability.
+ **Abstracted the Rasterizer**: Introduced an abstraction layer to enhance modularity.
+ **Introduced Uniform Buffer System**: Implemented a system for efficient management of shader uniforms.
+ **Modularized the Pipeline**: Aligned the rendering pipeline more closely with OpenGL's structure for improved programmability.
+ **Implemented Texture Mapping for Diffuse and Specular Components**: Enhanced visual realism by supporting texture mapping.
+ **Optimized Texture Loading**: Eliminated redundant texture loading to improve efficiency.
+ **Camera System**: Implement a simple and controllable camera system.
+ **Swap Buffer and Real-time Display**:  Introduced a swap buffer system to facilitate real-time rendering and display. By implementing double buffering, the rendering pipeline now supports smooth frame transitions and reduces visual artifacts such as flickering. 

Screenshot 2024-09-14 at 20 04 55

## Test Platform

macOS 14.5
Linux (Ubuntu 20.04)

## TODO:
+ Optimize performance for smoother real-time rendering.
+ Implement normal mapping.
+ Update comments, documentation and README.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -45,6 +45,8 @@ set(LOG_FILE_PATH "${EXECUTABLE_OUTPUT_PATH}/logs/SimpleRendererLog.log")
 set(LOG_FILE_MAX_SIZE 1024*1024*4)
 # 日志文件数量
 set(LOG_FILE_MAX_COUNT 8)
+#Debug Type (for cmake debug)
+set(CMAKE_BUILD_TYPE "Debug")
```
Comment (MRNIU) [inline CMakeLists.txt]: 这个改到 CMakePresets.json 吧

File: src/include/buffer.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    return drawBuffer_[index];
+  }
+
+ private:
+  size_t width_;
+  size_t height_;
+  size_t size_;
+  std::unique_ptr<uint32_t[]> framebuffer_1_;
+  std::unique_ptr<uint32_t[]> framebuffer_2_;
```
Comment (MRNIU) [inline src/include/buffer.hpp:69]: 这个能不能写成 vector 或者 array 的形式，方便拓展

File: test/system_test/camera.h [hunk 3]
```diff
@@ -25,33 +25,66 @@
...
+  glm::vec3 GetUp() const { return up_; }
+
+  // Methods to get view and projection matrices
+  glm::mat4 GetViewMatrix() const {
+    return glm::lookAt(position_, position_ + front_, up_);
+  }
+
+  glm::mat4 GetProjectionMatrix(float fov, float aspectRatio, float nearPlane,
+                                float farPlane) const {
+    return glm::perspective(glm::radians(fov), aspectRatio, nearPlane,
```
Comment (MRNIU) [inline test/system_test/camera.h:84]: 这个和 lookAt 我们把算法暴露出来吧，这也算是个知识点

File: test/system_test/camera.h [hunk 4]
```diff
@@ -25,33 +25,66 @@
...
+    position_ += glm::normalize(glm::cross(front_, up_)) * distance;
+  }
+
+  void MoveUp(float distance) { position_ += up_ * distance; }
+
+  void Rotate(float yaw, float pitch) {
+    glm::vec3 newFront;
+    newFront.x = cos(glm::radians(yaw)) * cos(glm::radians(pitch));
+    newFront.y = sin(glm::radians(pitch));
+    newFront.z = sin(glm::radians(yaw)) * cos(glm::radians(pitch));
```
Comment (MRNIU) [inline test/system_test/camera.h]: sin cos 加上 std::

File: src/material.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
+#include "material.hpp"
+
+#include <string>
+
+#define STB_IMAGE_IMPLEMENTATION
+#include "log_system.h"
+#include "stb_image.h"
+
+namespace simple_renderer {
+Texture Texture::LoadTextureFromFile(const std::string& path) {
```
Comment (MRNIU) [inline src/material.cpp:10]: 不是有构造析构吗，为啥还要专门实现一个方法

File: src/include/buffer.hpp (hunk 2 above)
Comment (ZzzhHe) [reply src/include/buffer.hpp:69]: 是要把uint32_t[]替换为vector吗？因为我感觉uint32_t[]更加直观用起来也简单

File: src/material.cpp (hunk 5 above)
Comment (ZzzhHe) [reply src/material.cpp:10]: 因为这个函数是个static函数，直接从文件中加载材质，然后返回一个材质。到处都可以直接用，不需要专门创建一个Texture对象。

**Pull request merged**